odoo/addons/base/models/res_users.py

```diff
@@ -199,6 +199,85 @@ def _search(self, args, offset=0, limit=None, order=None, count=False, access_ri
+            return {}
+        inherited_groups = {}
+        self = self._origin
+        for selected_group in self:
+            # Prepare dictionary, with key = selected group, values = all the groups that matches their
+            # implied groups' category with selected group's category
+            inherited_groups[selected_group] = self.filtered(
+                lambda group:
+                selected_group.category_id in group.implied_ids.category_id and
+                selected_group.category_id != group.category_id and
+                selected_group not in group.implied_ids
+            )
+        warnings = []
+        for selected_group, other_groups in inherited_groups.items():
+            if other_groups:
+                # From the groups user has, find the ones with selected group's category
+                selected_category_groups = self.filtered(lambda group: group.category_id == selected_group.category_id)
+                # Get all the implied groups of a selected group
+                selected_category_implied_groups = selected_category_groups._get_recursive_implied_groups(selected_category_groups)
+                # Get the minimum group needed in selected group's category (For example, someone changes
+                # Sales: Admin to Sales: User, but Field Service is already set to Admin, so here in the
+                # 'Sales' category, we will at the minimum need Admin group)
+                minimum_required_groups = other_groups.implied_ids.filtered(
+                    lambda group:
+                    group.category_id == selected_group.category_id and
+                    group not in selected_category_implied_groups
+                )
+                if minimum_required_groups:
+                    all_selected_category_groups = self.search([('category_id', '=', selected_group.category_id.id)])
+                    # Groups that have higher access level than minimum_required_groups
+                    possible_groups = all_selected_category_groups - minimum_required_groups._get_recursive_implied_groups(self.browse())
+                    warnings.append(_(
+                        "Since %s is a/an %s %s, you can set %s right only to %s"
+                        ) % (
+                            user.name,
+                            other_groups[0].category_id.name,
+                            other_groups[0].name,
+                            selected_group.category_id.name,
+                            _(" or ").join(possible_groups.mapped('name'))
+                        )
+                    )
+        return warnings
+
+    def _prepare_warnings_for_removed_groups(self, user, categories):
+        warnings = []
+        for category in categories:
+            # Find the group that enforces that at-least 'minimum_required_groups' in empty categories
+            enforcing_groups = self.filtered(lambda grp: category in grp.implied_ids.category_id)
+            if enforcing_groups:
```

Comment:
(same idea as above, so we don't indent when it's not needed)
```suggestion
            if not enforcing_groups:
                continue
            ...
```